### PR TITLE
Enhancements

### PR DESCRIPTION
**Pull request by pedromrb**
Description: 

Added a decorator function, 
Added a set functionality with the ability to throw Typerror exception, 
Added (as proposal) a readonly optional argument if we want to simulate partial immutability

Any doubts/concerns I'm available,

Thank you

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,14 +1,29 @@
...
 
 			return Reflect.get(target, name, receiver);
+		},
+
+		set: function(target, name, val, receiver) {
+			if (!(name in target) && readonly) {
+				throw new Error('Readonly object: Please don\'t set properties on this object.');
+			}
+
+			target[name] = val;
```
Comment (satazor) [inline index.js]: I think we must use Reflect.set here, see https://github.com/tvcutsem/harmony-reflect/blob/master/doc/api.md#reflectsettarget-name-value-receiver

File: index.js [hunk 2]
```diff
@@ -1,14 +1,29 @@
 'use strict';
+
 require('harmony-reflect');
 
-module.exports = function (obj) {
-	return new Proxy(obj, {
+module.exports = function (target, readonly) {
```
Comment (satazor) [inline index.js]: I think that we should keep the current API and add the decorator on top, either by detecting if `target` is a constructor or an object or adding a .decorator() method.

Comment (sindresorhus) [inline index.js]: I would add a `.decorator()` method.

File: readme.md [hunk 3]
```diff
@@ -26,6 +26,26 @@ obj = known(obj);
 // throws a TypeError when you try to access an unknown property
 console.log(obj.bar);
 //=> [TypeError] Unknown property: bar
+
+// setting a property, not much difference from actually setting but if not able, now it throw a TypeError exception
```
Comment (sindresorhus) [inline readme.md]: This comment could use some more work.

File: readme.md [hunk 4]
```diff
@@ -26,6 +26,26 @@ obj = known(obj);
...
+try {
+  obj.randomProperty = 'random';
+} catch (error) {
+  console.log(error instanceof TypeError);
+  console.error(error.message);
+}
+
+// setting readonly flag to true
+obj = known(obj, true);
```
Comment (sindresorhus) [inline readme.md]: I don't like boolean traps. Let's go with an options object instead:

```js
obj = known(obj, {readOnly: true});
```

Note the casing of the name.

File: readme.md [hunk 5]
```diff
@@ -26,6 +26,26 @@ obj = known(obj);
...
+
+// setting a property, not much difference from actually setting but if not able, now it throw a TypeError exception
+try {
+  obj.randomProperty = 'random';
+} catch (error) {
+  console.log(error instanceof TypeError);
+  console.error(error.message);
+}
+
+// setting readonly flag to true
```
Comment (sindresorhus) [inline readme.md]: Comments should explain how or why, not what.

File: index.js [hunk 6]
```diff
@@ -1,14 +1,29 @@
...
+			target[name] = val;
+
+			return true;
 		}
-	});
+	};
+
+	return function () {
+		return new Proxy(target, handler);
+	}.call(this, target);
```
Comment (sindresorhus) [inline index.js]: What's the purpose of this?

**Comment by sindresorhus** [comment]
Can you split this up? It's really two pull requests in one. The main request was the decorator. The readOnly option is unrelated.

File: index.js (hunk 6 above)
Comment (satazor) [inline index.js]: Yea .call is unnecessary.

File: index.js (hunk 2 above)
Comment (pedromrb) [inline index.js]: Yes, agree, wasnt sure on which direction to take, also thought in invoking like this:

decorator(know(target)) 

but like this is better, on a side node, i called the method decorateWith(arg), seemed more semantic

File: index.js (hunk 1 above)
Comment (pedromrb) [inline index.js]: Will add it, with the next pr

File: index.js (hunk 6 above)
Comment (pedromrb) [inline index.js]: It was an error, left over code from testing like this: decorator(know(target))

**Comment by pedromrb** [comment]
Yes, If this gets merged i'll create a pr with the readonly and the comments addressed.

File: readme.md [hunk 7]
```diff
@@ -26,6 +26,7 @@ obj = known(obj);
 // throws a TypeError when you try to access an unknown property
 console.log(obj.bar);
 //=> [TypeError] Unknown property: bar
+
 ```
```
Comment (sindresorhus) [inline readme.md]: Need to show a usage example using syntax from the decorator proposal.

File: test.js [hunk 8]
```diff
@@ -1,12 +1,68 @@
 'use strict';
+
 var assert = require('assert');
+var chai = require('chai');
```
Comment (sindresorhus) [inline test.js]: I prefer to just use normal `assert`. No benefit in using `chai` here.

File: test.js [hunk 9]
```diff
@@ -1,12 +1,68 @@
 'use strict';
+
```
Comment (sindresorhus) [inline test.js]: Don't do unrelated changes.

File: index.js (hunk 2 above)
Comment (sindresorhus) [inline index.js]: > but like this is better, on a side node, i called the method decorateWith(arg), seemed more semantic

That doesn't make much sense when you think about how it's going to be used:

```js
@known.decorateWith
class Foobar {
    baz = 1;
    someMethod() {}
}
```

This looks better:

```js
@known.decorator
class Foobar {
    baz = 1;
    someMethod() {}
}
```

**Comment by sindresorhus** [comment]
Needs to be documented in the readme. Reference the decorator spec. And give the pull request a more descriptive title. You should also have `Fixes #1` in the PR body, so the issue is closed when this PR is merged.

File: index.js [hunk 10]
```diff
@@ -1,14 +1,23 @@
...
 
-			return Reflect.get(target, name, receiver);
+			return Reflect.get(target, name);
 		}
-	});
+	};
+
+	var instance = new Proxy(target, handler);
+
+	instance.decorateWith = function (fn) {
```
Comment (sindresorhus) [inline index.js]: This shouldn't be added to the instance, but rather be a separate export. Please read the spec on how decorators work.

It should be used like this on a plain class:

```js
@known.decorator
class Foobar {
    baz = 1;
    someMethod() {}
}
```

Comment (pedromrb) [inline index.js]: Can you provide some insight on how to use the @sintax on a context where no transpiling occurs?

I'm trying to follow something like [this](https://addyosmani.com/blog/decorator-pattern/)

If the Proxy hijacks the property definition, I'm not sure on how an export can be combined with the known instantiation to expose the decorator method.

Comment (sindresorhus) [inline index.js]: > Can you provide some insight on how to use the @sintax on a context where no transpiling occurs?

It's explained in the spec: 

> A decorator is an expression, that evaluates to a function, that takes the target, name, and decorator descriptor as arguments.

So it's just a simple function that accepts the class and more and does something to it.

Some more examples here: http://babeljs.io/docs/plugins/transform-decorators/

> I'm trying to follow something like this

This is not the same thing.

Comment (pedromrb) [inline index.js]: Maybe I'm missing the point here, but the difficulties are a bit connected to this [discussion](https://github.com/wycats/javascript-decorators/issues/7).

If i try to do something like this and export it 
```javascript 
function isTestable(value) {
   return function decorator(target) {
      target.isTestable = value;
   }
}
```
i'll end up with an.. say ```export.decorate = function...etc``` and the calling of known (correct me if i'm wrong) will prevail with ```module.export``` or if i turn the export into an object and expose both methods the invoking will not be the same.

If these concerns are not correct or not applicable i'll drop the PR and maybe someone can do the proper implementation requested of this

Comment (satazor) [inline index.js]: @pedromrb That discussion has to do with how to use the decorator, which doesn't apply here.. we are just offering a decorator. People might use the decorator as a normal javascript function or with `@`.. it really is up to them. For instance, in our projects at moxy, we do not use decorators yet because they were not approved by TC39. Instead we use them using "vanilla" javascript

```
export default known.decorator()(myObject);
```

@sindresorhus one thing though, is that #1 is related to classes, not plain objects. This means that `known` must be applied to the instance and not to the constructor/proptotype properties. What do you suggest?

Comment (satazor) [inline index.js]: What I'm saying is that the usage that we are implementing here, will only lock down "static" properties of the class. I don't know if #1 wants that or instance properties.

Comment (pedromrb) [inline index.js]: Basically i tried several "workarounds" trying to not perturb the default invocation. 
@satazor yes, when i tried a similar solution to your suggestion when, for example extending an default property onto the object with a decorator, the Proxy was "hijacking" (this is why i added the decorator exposure to the instance, to bypass a situation like this).
